libgadget/cooling_qso_lightup.c

```diff
@@ -0,0 +1,622 @@
+    LMFP_intp = gsl_interp_alloc(gsl_interp_linear,Nreionhist);
+    gsl_interp_init(HeIII_intp, He_zz, XHeIII, Nreionhist);
+    gsl_interp_init(LMFP_intp, He_zz, LMFP, Nreionhist);
+
+    if(QSOLightupParams.heIIIreion_start > 1/He_zz[0]-1)
+        endrun(2, "HeII: Reionization table starts at z=%g but reionization starts earlier at z=%g\n", 1/He_zz[0]-1, QSOLightupParams.heIIIreion_start);
+
+    message(0, "Read %d lines z = %g - %g from file %s\n", Nreionhist, 1/He_zz[0] -1, 1/He_zz[Nreionhist-1]-1, reion_hist_file);
+}
+
+void
+init_qso_lightup(char * reion_hist_file)
+{
+    if(QSOLightupParams.QSOLightupOn)
+        load_heii_reion_hist(reion_hist_file);
+}
+
+static double last_zz;
+static double last_long_mfp_heating;
+
+/* Get the long mean free path heating. in erg/s/cm^3 */
+double
+get_long_mean_free_path_heating(double redshift)
+{
+    if(!QSOLightupParams.QSOLightupOn)
+        return 0;
+    if(redshift > QSOLightupParams.heIIIreion_start)
+        return 0;
+    if(redshift == last_zz)
+        return last_long_mfp_heating;
+    double atime = 1/(1+redshift);
+
+    /* Guard against the end of the table*/
+    if(atime > He_zz[Nreionhist-1])
+        return 0;
+
+    double long_mfp_heating = gsl_interp_eval(LMFP_intp, He_zz, LMFP, atime, NULL);
+
+    last_zz = redshift;
+    last_long_mfp_heating = long_mfp_heating;
+    return long_mfp_heating;
+}
+
+/* This function gets a random number from a Gaussian distribution using the Box-Muller transform.*/
+static double gaussian_rng(double mu, double sigma, const int64_t seed)
+{
+    double u1 = get_random_number(seed);
+    double u2 = get_random_number(seed + 1);
+    double z1 = sqrt(-2 * log(u1) ) * cos(2 * M_PI * u2);
+    return mu + sigma * z1;
+}
+
+/* Build a list of halos which are candidates for becoming a quasar.
+ * We use only halos with the right mass range.*/
+static int
+build_qso_candidate_list(int ** qso_cand, FOFGroups * fof)
+{
+    /*Loop over all halos, building the candidate list.*/
+    int i, ncand=0;
+    *qso_cand = mymalloc("Quasar_candidates", sizeof(int) * (fof->Ngroups+1));
+    for(i = 0; i < fof->Ngroups; i++)
+    {
+        /* Check that it has the right mass*/
+        if(fof->Group[i].Mass < QSOLightupParams.qso_candidate_min_mass)
+            continue;
+        if(fof->Group[i].Mass > QSOLightupParams.qso_candidate_max_mass)
+            continue;
+        /*Add to the candidate list*/
+        (*qso_cand)[ncand] = i;
+        ncand++;
+    }
+    /*Poison value at the end for safety.*/
+    (*qso_cand)[ncand] = -1;
+    *qso_cand = myrealloc(*qso_cand, (ncand+1) * sizeof(int));
+    return ncand;
+}
+
+/* Choose a black hole particle at random to host a quasar from the list of black hole particles.
+ * This function chooses a single quasar from the total candidate list of quasars on *all* processors.
+ * We seed the random number generator off the number of existing quasars.
```

Comment:
add 'Returns the local index of the halo in fof' if the halo is hosted on this rank, -1 if the halo is not hosted on this rank', or something alike.